pkg/asset/agent/iso.go

```diff
@@ -0,0 +1,92 @@
+package agent
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/openshift/assisted-image-service/pkg/isoeditor"
+	"github.com/openshift/installer/pkg/agent/isosource"
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/pkg/errors"
+)
+
+const (
+	agentISOFilename = "output/agent-installer.iso"
+)
+
+// ISO is an asset that generates the bootable image used to install clusters.
+type ISO struct {
+	File *asset.File
+}
+
+var _ asset.WritableAsset = (*ISO)(nil)
+
+// Dependencies returns the assets on which the Bootstrap asset depends.
+func (a *ISO) Dependencies() []asset.Asset {
+	return []asset.Asset{
+		&Ignition{},
+	}
+}
+
+// Generate generates the image file for to ISO asset.
+func (a *ISO) Generate(dependencies asset.Parents) error {
+	ignition := &Ignition{}
+	dependencies.Get(ignition)
+
+	// TODO: Use BaseIso asset once it is available
+	baseImage, err := isosource.EnsureIso()
+	if err != nil {
+		return err
+	}
+
+	ignitionByte, err := json.Marshal(ignition.Config)
+	if err != nil {
+		return err
+	}
+
+	ignitionContent := &isoeditor.IgnitionContent{Config: ignitionByte}
+	custom, err := isoeditor.NewRHCOSStreamReader(baseImage, ignitionContent, nil)
+	if err != nil {
+		return err
+	}
+	defer custom.Close()
+
+	output, err := os.Create(agentISOFilename)
+	if err != nil {
+		return err
+	}
+	defer output.Close()
+
+	_, err = io.Copy(output, custom)
+	return err
+}
+
+// Name returns the human-friendly name of the asset.
+func (a *ISO) Name() string {
+	return "Agent Installer ISO"
+}
+
+// Load returns the ISO from disk.
+func (i *ISO) Load(f asset.FileFetcher) (bool, error) {
+
+	iso, err := f.FetchByName(agentISOFilename)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, errors.Wrap(err, fmt.Sprintf("failed to load %s file", agentISOFilename))
+	}
+
+	i.File = iso
+	return true, nil
```

Comment:
There's probably no point implementing `Load()` - nothing depends on this, so we are never going to read an ISO loaded from disk, and in any event `Generate()` doesn't produce a file returnable from `Files()`. We're really only implementing this because the target has to take a list of WritableAssets, not any Assets. It's safe to always just `return false, nil`.